Code/src/main/java/nl/utwente/group10/haskell/expr/Function.java

```diff
@@ -0,0 +1,124 @@
+package nl.utwente.group10.haskell.expr;
+
+import nl.utwente.group10.haskell.env.Env;
+import nl.utwente.group10.haskell.exceptions.HaskellException;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
+import nl.utwente.group10.haskell.type.FuncT;
+import nl.utwente.group10.haskell.type.Type;
+
+import java.util.*;
+
+/**
+ * Represents a user-defined Haskell function. Holds arguments which can be used for building the function.
+ */
+public class Function extends Expr {
+
+    /**
+     * An argument for a function. A FunctionArgument can be used as a variable in a function definition.
+     */
+    public static class FunctionArgument extends Expr {
+        /** The type for this argument. */
+        private Type type;
+
+        /** The name of this argument. Will be randomly generated to a unique value. */
+        private String name;
+
+        /**
+         * Constructor for a FunctionArgument with an explicit type.
+         *
+         * @param type The required type for this argument.
+         */
+        public FunctionArgument(Type type) {
+            this.type = type;
+            this.name = String.format("var_%s", UUID.randomUUID().toString().substring(0, 8));
+        }
+
+        @Override
+        public Type analyze(Env env, GenSet genSet) throws HaskellException {
+            return this.type;
+        }
+
+        @Override
+        public String toHaskell() {
+            return this.name;
+        }
+
+        @Override
+        public String toString() {
+            return this.name;
+        }
+    }
+
+    /** The arguments for this function. */
+    private List<FunctionArgument> arguments;
+
+    /** The expression that forms the base of this function. */
+    private Expr expr;
+
+    /**
+     * @param expr The expression that forms the base of this function.
+     */
+    public Function(Expr expr, FunctionArgument ... arguments) {
+        this.arguments = Arrays.asList(arguments);
+        this.expr = expr;
+    }
+
+    /**
+     * @return An ordered array of the arguments for this function.
+     */
+    public FunctionArgument[] getArguments() {
```

Comment:
Because of the ```toArray``` a new object is returned - so there is no way to change the ```Function```, only the elements in it (which should be immutable aside from ```Value```).